src/qttools/obc/spectral.py

```diff
@@ -0,0 +1,290 @@
+        dEk_dk = xp.zeros_like(ws)
+
+        # NOTE: This is actually only correct if we have no overlap.
+        # phi.H d/dk (H00 + lambda * H01 + 1/lambda * H10) phi =
+        # phi.H d/dk energy (S00 + lambda * S01 + 1/lambda * S10) phi
+        with warnings.catch_warnings(
+            action="ignore", category=RuntimeWarning
+        ):  # Ignore division by zero.
+            for i in range(batchsize):
+                for j, w in enumerate(ws[i]):
+                    a = -(1j * w * a_ij[i] - 1j / w * a_ji[i])
+                    phi = vs[i, :, j]
+                    phi_t = phi.conj().T
+                    dEk_dk[i, j] = (phi_t @ a @ phi) / (phi_t @ phi)
+
+            ks = -1j * xp.log(ws)
+
+        # Find eigenvalues that correspond to reflected modes. These are
+        # modes that either propagate into the leads or decay away from
+        # the system.
+        return ((dEk_dk.real < 0) & (xp.abs(ks.imag) < self.min_decay)) | (
+            (ks.imag < -self.min_decay) & (ks.imag > -self.max_decay)
+        )
+
+    def _compute_x_ii(
+        self,
+        a_ii: xp.ndarray,
+        a_ij: xp.ndarray,
+        a_ji: xp.ndarray,
+        ws: xp.ndarray,
+        vs: xp.ndarray,
+        mask: xp.ndarray,
+    ) -> xp.ndarray:
+        """Computes the surface Green's function.
+
+        Parameters
+        ----------
+        a_ii : xp.ndarray
+            The diagonal block of the periodic matrix.
+        a_ij : xp.ndarray
+            The superdiagonal block of the periodic matrix.
+        a_ji : xp.ndarray
+            The subdiagonal block of the periodic matrix.
+        ws : xp.ndarray
+            The eigenvalues of the NEVP.
+        vs : xp.ndarray
+            The eigenvectors of the NEVP.
+        mask : xp.ndarray
+            A boolean mask indicating which eigenvalues correspond to
+            reflected modes.
+
+        Returns
+        -------
+        x_ii : xp.ndarray
+            The surface Green's function.
+
+        """
+        if self.x_ii_formula == "self-energy":
+            # Equation (13.1).
+            x_ii_a_ij = xp.zeros((mask.shape[0], *a_ij.shape[-2:]), dtype=a_ij.dtype)
+            for i, m in enumerate(mask):
+                vs_ = vs[i][:, m]
+                # Moore-Penrose pseudoinverse.
+                vs_inv = xp.linalg.inv(vs_.T @ vs_) @ vs_.T
+                x_ii_a_ij[i] = vs_ / ws[i, m] @ vs_inv
+
+            # Calculate the surface Green's function.
+            return xp.linalg.inv(a_ii + a_ji @ x_ii_a_ij)
+
+        if self.x_ii_formula == "direct":
+            # Equation (13.2).
+            x_ii = xp.zeros((mask.shape[0], *a_ij.shape[-2:]), dtype=a_ij.dtype)
+            for i, m in enumerate(mask):
+                v = vs[i][:, m]
+                w = ws[i, m]
+                # Direct computation of the surface Green's function.
+                inverse = npla.inv(v.conj().T @ a_ij[i] @ v * w)
```

Comment:
x_ii is xp.zeros but the inverse is calling npla (that if I'm not wrong is NumPy), shouldn't it be xp.linalg? So it can invert on GPU too?

Reply:
Yes, again something i overlooked. I actually also need to add tests that the different formulas work as expected.

Reply:
Sounds good